src/common/middleware/call-api.js

```diff
@@ -17,39 +17,67 @@ export default defaults => () => next => async action => {
   validateDefaults(defaults);
   validateSettings(settings);
 
-  const [ requestType, successType, failureType ] = settings.types;
+  const requestType = settings.types ? settings.types[0] : null;
+  const successType = settings.types ? settings.types[1] : null;
+  const failureType = settings.types ? settings.types[2] : null;
   const resource = defaults.endpoint + settings.path;
   const optionsWithCsrfToken = withCsrfToken(defaults.csrfToken, settings.options);
   const createAction = createActionWith.bind({}, action);
 
   // Immediately dispatch request action
-  next(createAction({ type: requestType }));
+  if (requestType) {
+    next(createAction({ type: requestType }));
+  }
 
-  try {
-    const response = await fetch(resource, optionsWithCsrfToken);
-    await checkStatus(response);
-    const result = await response.json();
+  return fetch(resource, optionsWithCsrfToken)
+    .then((response) => {
+      return checkStatus(response).then(() => {
+        return response.json().then((result) => {
+          if (result.status !== 'success') {
+            throw getError(response, result);
+          }
 
-    if (result.status !== 'success') {
-      throw getError(response, result);
-    }
+          if (successType) {
+            return next(createAction({
+              type: successType,
+              payload: {
+                response: result
+              }
+            }));
+          }
 
-    return next(createAction({
-      type: successType,
-      payload: {
-        response: result
-      }
-    }));
-  }
-  catch(error) {
-    return next(createAction({
-      type: failureType,
-      payload: {
-        error: error
-      },
-      error: true
-    }));
-  }
+          return Promise.resolve(result);
+        });
+      }, (error) => {
+        if (failureType) {
+          return next(createAction({
+            type: failureType,
+            payload: {
+              error: error
+            },
+            error: true
+          }));
+        }
+
+        error.action = action;
+
+        return Promise.reject(error);
+      }, (error) => {
+        if (failureType) {
+          return next(createAction({
+            type: failureType,
+            payload: {
+              error: error
+            },
+            error: true
+          }));
+        }
+
+        error.action = action;
+
+        return Promise.reject(error);
```

Comment:
It would be good to have this discussion in bigger group to find a good compromise.

Creating a webhook is just a one quite simple example, because at this point we don't even do much with the result, we just want to wait till request is finished before proceeding.

We have much more complicated examples around store auth and name registering, where actions involve calling multiple APIs for authorization, exchanging tokens, etc. Trying to make them all fit into CALL_API middleware will likely make middleware full of special case 'workarounds' - like the one with returning a Promise here.

Returning a Promise from the middleware seems a bit inconsistent to me, especially that we can't rely on dispatch ALWAYS returning a promise, because promise will only be returned for if CALL_API finds settings object, not for any other actions.
If we end up having middleware returning promises from dispatch, I would expect that any dispatch call on any action would return a promise, so we would be able to consistently rely on it, not thinking if this particular action goes through the middleware or not.

At this point - before we have the discussion and find a better solution for it, I would suggest we only use CALL_API middleware for API calls that result in redux actions.
Yes, it means that not all of our API calls will go through middleware, but as far as I can tell, the ones that currently don't fit into CALL_API scheme don't really need CSRF tokens (for example store API calls that we intend to do directly to store once it allows cross origin requests).